File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
    "math"
    "math/rand"
    "sort"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
    modifier := math.Floor(float64(score-10) / 2)
	return int(modifier)
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	rolls := make([]int, 0, 4)
    for i := 0; i < 4; i++ {
        roll := rand.Intn(6) + 1
        rolls = append(rolls, roll)
    }
    sort.Ints(rolls)
    return rolls[1] + rolls[2] + rolls[3]
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	char := Character{
        Strength:     Ability(),
    	Dexterity:    Ability(),
    	Constitution: Ability(),
    	Intelligence: Ability(),
    	Wisdom:       Ability(),
    	Charisma:     Ability(),
    }
    char.Hitpoints = 10 + Modifier(char.Constitution)
    return char
}

File: solutions/csharp/dnd-character/1/DndCharacter.cs
public class DndCharacter
{
    private static Random random = new Random();
    public int Strength { get; set; }
    public int Dexterity { get; set; }
    public int Constitution { get; set; }
    public int Intelligence { get; set; }
    public int Wisdom { get; set; }
    public int Charisma { get; set; }
    public int Hitpoints { get; set; }

    public static int Modifier(int score) =>
        (int)Math.Floor((score - 10) / 2.0);

    public static int Ability() 
    {
        List<int> rolls = new List<int>();
        for (int i = 0; i < 4; i++)
        {
            int roll = random.Next(1,7);
            rolls.Add(roll);
        }
        rolls.Sort();
        return rolls.Sum() - rolls[0]; 
    }

    public static DndCharacter Generate()
    {
        var character = new DndCharacter
        {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = Ability(),
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability()
        };
        character.Hitpoints = 10 + Modifier(character.Constitution);
        return character;
    }
}
